src/main/scala/firrtl/passes/Passes.scala

```diff
@@ -247,23 +249,204 @@ object Legalize extends Pass {
+  * @define implicitScope @param scope the enclosing scope of this name. If [[None]], then this is a [[Circuit]] name
+  */
+object VerilogRename extends Transform {
+  def inputForm: CircuitForm = LowForm
+  def outputForm: CircuitForm = LowForm
+  type Renames = mutable.HashMap[String, String]
+  private val inlineDelim = "_"
+
+  /** Generate a new name, by appending underscores, that will not conflict with the existing namespace
+    * @param n a name
+    * @param ns a [[Namespace]]
+    * @return a conflict-free name
+    * @note prefix uniqueness is not respected
+    */
+  private def safeName(n: String, ns: Namespace): String =
+    Uniquify.findValidPrefix(n + inlineDelim, Seq(""), ns.cloneUnderlying ++ v_keywords)
+
+  /** Modify a name to not conflict with a Verilog keywords while respecting existing renames and a namespace
+    * @param n the name to rename
+    * @param renames the [[RenameMap]] to query when renaming
+    * $implicitRename
+    * $implicitNamespace
+    * $implicitScope
+    * @return a name without keyword conflicts
+    */
+  private def onName(n: String)(implicit renames: RenameMap, ns: Namespace, scope: Option[Named]): String = {
+
+    // Convert a [[String]] into [[Named]] based on the provided scope.
+    def wrap(name: String, scope: Option[Named]): Named = scope match {
+      case None                     => CircuitName(name)
+      case Some(cir: CircuitName)   => ModuleName(name, cir)
+      case Some(mod: ModuleName)    => ComponentName(name, mod)
+      case Some(com: ComponentName) => ComponentName(s"${com.name}.$name", com.module)
+    }
+
+    val named = wrap(n, scope)
+
+    // If this has already been renamed use that name. If it conflicts with a keyword, determine a new, safe name and
+    // update the renames. Otherwise, leave it alone.
+    val namedx: Seq[Named] = renames.get(named) match {
+      case Some(x) => x
+      case None if (v_keywords(n)) =>
+        val sn = wrap(safeName(n, ns), scope)
+        renames.rename(named, sn)
+        Seq(sn)
+      case _ => Seq(wrap(n, scope))
+    }
+
+    namedx match {
+      case (ComponentName(n, _) :: Nil) => n
+      case (ModuleName(n, _)    :: Nil) => n
+      case (CircuitName(n)      :: Nil) => n
+      case x => throw new PassException(
+        s"Verilog renaming shouldn't result in multiple renames, but found '$named -> $namedx'")
+    }
+  }
+
+  /** Rename the fields of a [[Type]] to match the ports of an instance
+    * @param t the type to rename
+    * $implicitRename
+    * $implicitNamespace
+    * $implicitScope
+    * @return a [[Type]] with updated names
+    * @note This is not intended for fixing arbitrary types, only [[BundleType]] in instance [[WRef]]s
+    */
+  private def onType(t: Type)
+                    (implicit renames: RenameMap, ns: Namespace, scope: Option[ModuleName]): Type = t match {
+    case b: BundleType => b.copy(fields = b.fields.map(f => f.copy(name = onName(f.name))))
+    case t => t
```

Comment:
suspicious shadowing, this is a trivial case but I believe firrtl convention is to go `tt => tt` here, or `_ => t`